api/app/src/main/kotlin/packit/service/RolePermissionService.kt

```diff
@@ -2,83 +2,182 @@ package packit.service
 
 import org.springframework.http.HttpStatus
 import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
 import packit.exceptions.PackitException
-import packit.model.Role
-import packit.model.RolePermission
+import packit.model.*
 import packit.model.dto.UpdateRolePermission
-import packit.repository.*
+import packit.repository.RoleRepository
 
 interface RolePermissionService
 {
-    fun removeRolePermissionsFromRole(role: Role, removeRolePermissions: List<UpdateRolePermission>): Role
-    fun getRolePermissionsToAdd(role: Role, addRolePermissions: List<UpdateRolePermission>): List<RolePermission>
+    fun updatePermissionsOnRole(
+        role: Role,
+        addRolePermissions: List<UpdateRolePermission>,
+        removeRolePermissions: List<UpdateRolePermission>,
+    ): Role
+
+    fun updatePermissionOnRoles(
+        rolesToAdd: List<Role>,
+        rolesToRemove: List<Role>,
+        permissionName: String,
+        packetGroupName: String,
+        packetId: String? = null,
+    )
 }
 
 @Service
 class BaseRolePermissionService(
-    private val permissionRepository: PermissionRepository,
-    private val packetRepository: PacketRepository,
-    private val packetGroupRepository: PacketGroupRepository,
-    private val tagRepository: TagRepository,
-    private val rolePermissionRepository: RolePermissionRepository
+    private val permissionService: PermissionService,
+    private val packetService: PacketService,
+    private val packetGroupService: PacketGroupService,
+    private val tagService: TagService,
+    private val roleRepository: RoleRepository,
 ) : RolePermissionService
 {
+    @Transactional(rollbackFor = [Exception::class])
+    override fun updatePermissionsOnRole(
+        role: Role,
+        addRolePermissions: List<UpdateRolePermission>,
+        removeRolePermissions: List<UpdateRolePermission>
+    ): Role
+    {
+        addPermissionsToRole(role, addRolePermissions)
+        removeRolePermissionsFromRole(role, removeRolePermissions)
+
+        return roleRepository.save(role)
+    }
+
+    @Transactional(rollbackFor = [Exception::class])
+    override fun updatePermissionOnRoles(
+        rolesToAdd: List<Role>,
+        rolesToRemove: List<Role>,
+        permissionName: String,
+        packetGroupName: String,
+        packetId: String?,
+    )
+    {
+        val permission = permissionService.getByName(permissionName)
+        val packet = packetId?.let { packetService.getPacket(it) }
+        val packetGroup = if (packetId == null)
+        {
+            packetGroupService.getPacketGroupByName(packetGroupName)
+        } else
+        {
+            null
+        }
```

Comment:
What happens here if the packet doesn't match the packetGroup name? I think we should check that and fail.

Reply:
good catch!